### PR TITLE
invite users after signing up during setup

### PR DESCRIPTION
**Pull request by josephwegner**
So this turned in to a much bigger PR than I was expecting. There's a lot of nuance to doing this form correctly, so I had to rethink some things.
## tl;dr
- moves skip link to below the submit button
- makes the submit button better represent form status
- posts notifications based on success/failure of notifications
- goes to the invite page after user creation
- actually sends invites!
### Navigation

First off, I had to get the signup form to take the user to the invite page, rather than going directly to admin. In order to do that, before signing the user in, I change an option telling Ghost to skip the session success handler. Then I can redirect manually from the signup page.

_as of right now, the user cannot go back to step 2 after entering step 3._
### Skip invitations

Prior to this PR, the submit button on the invitation page defaulted to being a skip button if there were no emails in the textbox. That turned out to be problamatic, because I wanted to pop error notifications when the user clicked that button. I also generally wanted that button to reflect the state of the form, which means it needed to be deactivated if the form was invalid.

The change I made was to move the skip functionality into some text that sits underneath the button. So, when you hit the invitations page, this is what you will see.

screen shot 2015-05-28 at 2 30 44 pm
### Live form status stuff

So now that button is able to represent some state of the form. The two things it attempts to communicate are:

**number of invites**  - this is reflected in the text of the button (ie: "Invite some users", "Invite 1 user", "Invite 2 users")
**email validitiy** - this is reflect by the color of the button - grey, invalid. green, valid. If you have three emails in there, and 1 of the 3 is invalid, your button will be grey and disabled. Clicking it will pop a notification saying which emails are invalid.  

Some screenshots:

screen shot 2015-05-28 at 2 30 57 pm

screen shot 2015-05-28 at 2 31 04 pm

screen shot 2015-05-28 at 2 31 15 pm
### Notifications

I mentioned above that there are error notifications prior to sending invites. I also added some notifications post-invite to indicate success/failure. You will get one notification for success (it indicates the number of invites sent), and one notification listing the failed invites (indicates how many, and which emails failed).

I don't have a screenshot of errored invites, but here's a successful one:

screen shot 2015-05-28 at 2 31 29 pm
### General logic stuffs
- dedupes emails
- blank lines won't be counted as emails in the form state
- separate XHR request sent for every invite - waits until all finish before moving on.
### Things I'm worried about, and you should look for in review
- This is my first real ember work. Some of it feels dirty, and in general I don't know that I'm doing things right. Please be ruthless in that portion of the review - that will be the best way for me to learn.
- I was seeing occasional slug collisions when creating the users, which would fail the invite process. I couldn't reliably reproduce it, and I'm not sure why it would happen.
- On my localhost, sending all the invite emails kind of takes awhile. That makes the form feel a little weird.
  - Perhaps the api should do the email sending async, and just return with the user object as soon as possible
  - I'm not sure the common practice, but it might not hurt to show some loading status on the form
- If the user leaves the page, and somehow comes back with a fresh session, they will be taken to the invite page but can't actually do anything. They're not authenticated, so they can't send invites.

### CODE REVIEW

**Comment by josephwegner** [comment]
I know we're having some travis issues... if that's a unique error, lemme know... hopefully not.

File: core/client/app/controllers/setup/three.js [hunk 1]
```diff
@@ -4,38 +4,126 @@ import ValidationEngine from 'ghost/mixins/validation-engine';
...
-        var user = this.get('numUsers') === 1 ? 'user' : 'users';
-        return this.get('numUsers') > 0 ?
-            'Invite ' + this.get('numUsers') + ' ' + user : 'I\'ll do this later, take me to my blog!';
+    buttonClass: Ember.computed('validateUsers', 'numUsers', function () {
+        if (this.get('validateUsers') === true && this.get('numUsers') > 0) {
+            return 'btn-green';
+        } else {
+            return 'btn-minor';
+        }
+        return this.get('validateUsers') === true ? 'btn-green' : 'btn-minor';
```
Comment (jaswilli) [inline core/client/app/controllers/setup/three.js]: Is this line (54) reachable?

Comment (josephwegner) [inline core/client/app/controllers/setup/three.js]: ah, no, I don't believe so. I'll remove it.

**Comment by josephwegner** [comment]
@jaswilli I removed the unreachable bit.

I can squash once this gets through review, and everything is +1'd

**Comment by ErisDS** [comment]
The travis issues are fixed now, so rebasing should result in a green light.

**Comment by josephwegner** [comment]
rebased. :four_leaf_clover:

**Comment by josephwegner** [comment]
ok, all clean!

**Comment by josephwegner** [comment]
any blockers from merging this?

File: core/client/app/controllers/setup/three.js [hunk 2]
```diff
@@ -2,40 +2,130 @@ import Ember from 'ember';
...
     }),
     actions: {
         invite: function () {
-            console.log('inviting', this.get('usersArray'));
+            var self = this,
+                validationErrors = this.get('validateUsers'),
+                users = this.get('usersArray'),
+                errorMessages,
+                notifications = this.get('notifications');
+            console.log('NOTIFICATIONS', notifications);
```
Comment (ErisDS) [inline core/client/app/controllers/setup/three.js]: something to remove when you squash ;)

File: core/client/app/templates/setup/three.hbs [hunk 3]
```diff
@@ -12,7 +12,8 @@
 sally.sanders@example.com" value=users}}
     </form>
 
-    <button {{action "signin"}} class="btn btn-default btn-lg btn-block {{buttonClass}}">
+    <button {{action 'invite'}} class="btn btn-default btn-lg btn-block {{buttonClass}}">
         {{buttonText}}
     </button>
+    <a {{action 'signin'}} href="#" class="gh-flow-skip">I'll do this later, take me to my blog!</a>
```
Comment (ErisDS) [inline core/client/app/templates/setup/three.hbs]: I'm not sure about this approach?

**Comment by ErisDS** [comment]
Would be good to get some input from @jaswilli or @novaugust on the Ember side of this.

File: core/client/app/templates/setup/three.hbs (hunk 3 above)
Comment (josephwegner) [inline core/client/app/templates/setup/three.hbs]: You mean on the ember side, or the UI (as in, you don't like a text link down there).

On the ember side, I guess that should just be a regular link now, since signin happened in the previous step.

Comment (ErisDS) [inline core/client/app/templates/setup/three.hbs]: Sorry, I didn't mean to be so vague. I thought you had 2 elements on the page and were swapping them based on state, which seemed odd. However, I now realise they're actually both on the page at the same time, which is definitely not the right thing. They should be the same button as they were before.

Comment (JohnONolan) [inline core/client/app/templates/setup/three.hbs]: If it can be done per the spec, that would be preferable. If that is going to be really annoying/difficult (@jaswilli confirm?)  -- this solution is an acceptable alternative :+1:

Comment (josephwegner) [inline core/client/app/templates/setup/three.hbs]: it's not hard - I can go back to the old spec. My thought with doing it this way is now, if the user quickly clicks the main CTA on the page, they won't automatically be taken off the invite page... That _should_ turn in to more invites, because the user has to make a more conscious decision to skip.

Also, having form status reflected in the button is nice.

But, if y'all prefer the old way, I'll go back.

Comment (JohnONolan) [inline core/client/app/templates/setup/three.hbs]: > That should turn in to more invites, because the user has to make a more conscious decision to skip.

A good point, let's give it a try :)

Comment (jaswilli) [inline core/client/app/templates/setup/three.hbs]: This can be switched to a simple link-to instead of a button with an action:

```handlebars
{{#link-to "posts" class="gh-flow-skip"}}
    I'll do this later, take me to my blog!
{{/link-to}}
```

File: core/client/app/controllers/setup/three.js [hunk 4]
```diff
@@ -2,40 +2,130 @@ import Ember from 'ember';
...
+                    switch (error.error) {
+                    case 'email':
+                        return {message: error.user + ' is not a valid email.'};
+                    }
+                });
+
+                notifications.showErrors(errorMessages);
+            }
         },
         signin: function () {
```
Comment (jaswilli) [inline core/client/app/controllers/setup/three.js]: Can go away.

File: core/client/app/controllers/setup/three.js [hunk 5]
```diff
@@ -2,40 +2,130 @@ import Ember from 'ember';
...
+                this.get('authorRole').then(function (authorRole) {
+                    Ember.RSVP.Promise.all(
+                        users.map(function (user) {
+                            var newUser = self.store.createRecord('user', {
+                                email: user,
+                                status: 'invited',
+                                role: authorRole
+                            });
+
+                            return newUser.save().then(function () {
```
Comment (jaswilli) [inline core/client/app/controllers/setup/three.js]: There should probably be some error handling here.  As it stands if this save fails there's no feedback shown on the page at all.  It just appears as though nothing has happened.  You can test it by trying to invite the email address you just signed up with.

File: core/client/app/controllers/setup/three.js [hunk 6]
```diff
@@ -2,40 +2,130 @@ import Ember from 'ember';
 import ValidationEngine from 'ghost/mixins/validation-engine';
 
 var SetupThreeController = Ember.Controller.extend(ValidationEngine, {
```
Comment (jaswilli) [inline core/client/app/controllers/setup/three.js]: As far as I can tell this controller doesn't need to use `ValidationEngine`, so it shouldn't need to be imported or mixed-in here.

File: core/client/app/controllers/setup/two.js [hunk 7]
```diff
@@ -55,9 +55,14 @@ export default Ember.Controller.extend(ValidationEngine, {
                         }]
                     }
                 }).then(function () {
+                    // Don't call the success handler, otherwise we will be redirected to admin
+                    self.controllerFor('application').set('skipAuthSuccessHandler', true);
```
Comment (jaswilli) [inline core/client/app/controllers/setup/two.js]: `controllerFor` is going away.  Instead we should inject the application controller and then reference it like a normal property:

Above, where `notifications` and `ghostPaths` are injected:
`application: Ember.inject.controller()`

And then this can be changed to:
`self.get('application').set(...);`

File: core/client/app/controllers/setup/two.js [hunk 8]
```diff
@@ -55,9 +55,14 @@ export default Ember.Controller.extend(ValidationEngine, {
...
+
                     self.get('session').authenticate('simple-auth-authenticator:oauth2-password-grant', {
                         identification: self.get('email'),
                         password: self.get('password')
+                    }).then(function () {
```
Comment (jaswilli) [inline core/client/app/controllers/setup/two.js]: Since this controller sticks around in memory after we move off the screen we should `self.set('password', '')` after successfully authenticating so we don't hold onto a plaintext copy of the password any longer than necessary.

**Comment by jaswilli** [comment]
@josephwegner Sorry for the delay.  I made a handful of comments on some minor issues.  After those are cleaned up and this is rebased it should be good to go.

**Comment by josephwegner** [comment]
@jaswilli @ErisDS @JohnONolan OK, all issues fixed, and rebased against master!

GTG?
